text/0812-tracked-built-ins.md

```diff
@@ -0,0 +1,287 @@
+arbitrary number of players, and keep track of the score for each player using
+tracked version of JavaScript [Map][global-objects-map].
+
+**Note:** In this and the following examples, I am using `<template>` from
+[RFC #779][rfc-0779].
+
+```js
+// app/components/scoreboard.js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedMap } from 'tracked-built-ins';
+
+const PLAYERS = [
+  { name: 'Zoey' },
+  { name: 'Tomster' },
+];
+
+export default class Scoreboard extends Component {
+  // Create a map from player -> score, with each player's score starting at 0
+  scores = new TrackedMap(PLAYERS.map(p => [p, 0]));
+
+  @action
+  incrementScore(player) {
+    let currentScore = this.scores.get(player);
+
+    this.scores.set(player, currentScore + 1);
+  }
+
+  <template>
+    {{#each-in this.scores |player score|}}
+      <div>
+        {{player.name}}: {{score}}
+    
+        <button type="button" {{on "click" (fn this.incrementScore player)}}>
+          +
+        </button>
+      </div>
+    {{/each-in}}
+  </template>
+}
+```
+
+Lets lake a look into another example involving a shopping cart, where we keep track
+of the items added to cart using tracked version of JavaScript's [Array][global-objects-array].
+
+```js
+// app/components/scoreboard.js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedArray } from 'tracked-built-ins';
+
+export default class Cart extends Component {
+  items = new TrackedArray([
+    { productName: 'Jeans' },
+    { productName: 'T-shirt' },
+  ]);
+
+  @action
+  addPants(item) {
+    this.items.push(item);
+  }
+
+  <template>
+    {{#each this.items as |item|}}
+      <div>
+        {{item.productName}}
+      </div>
+    {{/each}}
+  
+    <button type="button" {{on "click" (fn this.addPants @pants)}}>
+      Add Pants
+    </button>
+  </template>
+}
+```
+
+## Drawbacks
+
+> Why should we *not* do this? Please consider the impact on teaching Ember,
+on the integration of this feature with other existing and planned features,
+on the impact of the API churn on existing apps, etc.
+
+> There are tradeoffs to choosing any path, please attempt to identify them here.
+
+## Alternatives
+
+- `tracked-built-ins` can be added only to the `app` blueprint and not to the `addon` blueprint.
+
+- Do nothing and recommend using re-setting the property to track data structure changes.
+
+  The tradeoffs of this approach described in "Motivation" section of this RFC.
+
+- Don't add `tracked-built-ins` to the blueprint and only update Ember Guides.
+
+  Ember.js is advertised as "batteries included" framework with complete out-of-the-box experience.
+  Asking users to install additional dependencies to achieve relatively common problem contradicts
+  with that paradigm.
+
+## Unresolved questions
+
+- Should we introduce these at something like `@glimmer/tracking` instead?
```

Comment:
This'd be my preference.

Also, providing an optional [`deepTracked`](https://github.com/NullVoxPopuli/ember-deep-tracked) for :musical_note: folks that just don't care :musical_note:

Reply:
@NullVoxPopuli interesting note! I think it's probably fair to say that that definitely won't happen as part of *this* RFC. It may be worth adding, but needs its own RFC, because there are non-trivial design and *teaching* questions about what it looks like to include that. (My own take, not necessarily shared by anyone else on the Framework or TS core teams, is that using it is an anti-pattern outside of tests, and that there are generally much better ways of solving the needs people reach for that pattern for.)

Reply:
(I agree lol)

whenever I've need deep tracking in practice, I've always found it easier to recursively use the cache primitives, so I can select what is tracked as I recurse